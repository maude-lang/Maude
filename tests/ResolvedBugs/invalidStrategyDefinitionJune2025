#!/bin/sh

MAUDE_LIB=$srcdir/../../src/Main
export MAUDE_LIB

../../src/Main/maude \
  < $srcdir/invalidStrategyDefinitionJune2025.maude -no-banner -no-advise \
  > invalidStrategyDefinitionJune2025.out 2>&1

diff $srcdir/invalidStrategyDefinitionJune2025.expected invalidStrategyDefinitionJune2025.out > /dev/null 2>&1
