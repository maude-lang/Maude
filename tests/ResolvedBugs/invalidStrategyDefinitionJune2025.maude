***
*** Test for strategy to expression mapping that instantiate the
*** left-hand side of a strategy definition to a expression other
*** than a call (not valid), to a call (valid) and to a call with
*** less variables.
*** 

sth STRIV is
	including TRIV .
	strat st1 @ Elt .
	strat st2 @ Elt .
	strat st3 : Elt @ Elt .
endsth

smod WRONG{X :: STRIV} is
	sd st1 := idle .
	sd st2 := fail .
	csd st3(E:X$Elt) := idle if E:X$Elt = E:X$Elt .
endsm

smod SBOOL is
	protecting BOOL .
	strat sb2 @ Bool .
	strat sb3 : Bool @ Bool .
endsm

view Match from STRIV to SBOOL is
	sort Elt to Bool .
	strat st1 to expr match true .  *** yields ill-formed definition
	strat st2 to expr sb2 .  *** yields valid definition
	var X : Elt .
	strat st3(X) to expr sb3(true) .  *** yields unbound variables
endv

smod MAIN is
	including WRONG{Match} .
endsm

srew true using sb2 .
srew false using sb3(true) .
show sds .
