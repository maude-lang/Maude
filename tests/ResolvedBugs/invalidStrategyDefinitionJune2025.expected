Warning: <standard input>, line 35 (smod MAIN): left-hand side of strategy
    definition
  sd st1 := idle .
would be instantiated to match true by a strategy to expression mapping. The
    definition would be ill-formed and thus it cannot be used.
Warning: <standard input>, line 18 (smod WRONG): variable E:Bool is used before
    it is bound in strategy definition:
csd sb3(true) := idle if E:Bool = E:Bool .
==========================================
srewrite in MAIN : true using sb2 .

No solution.
rewrites: 0 in 0ms cpu (0ms real) (~ rewrites/second)
==========================================
srewrite in MAIN : false using sb3(true) .

No solution.
rewrites: 0 in 0ms cpu (0ms real) (~ rewrites/second)
sd sb2 := fail .
csd sb3(true) := idle if E:Bool = E:Bool .
Bye.
